### PR TITLE
fix(bar-label): take overflow into finding best placement for bar label

### PR DESCRIPTION
**Pull request by quanho**
Overflow property is used to ignore checking if a label is fit into a rectangle.
So we should take this into account of finding best placement for labels.

### CODE REVIEW

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js [hunk 1]
```diff
@@ -167,10 +167,10 @@ export function findBestPlacement({
     largest = !p || testBounds.height > largest.height ? testBounds : largest;
 
     if (orientation === 'v' && ((fitsHorizontally && testBounds.height >= measured.height)
-      || (!fitsHorizontally && testBounds.height >= measured.width && testBounds.width >= measured.height))) {
+      || (!fitsHorizontally && testBounds.height >= measured.width && (!!placement.overflow || testBounds.width >= measured.height)))) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: Let's not write nested if statements like this. Break each condition down into its own statement similar to:

```js
if (orientation === 'v') {
      if (fitsHorizontally) {
        if (testBounds.height >= measured.height) {
          bounds = testBounds;
          break;
        }
      } else if (testBounds.height >= measured.width && testBounds.width >= measured.height) {
        bounds = testBounds;
        break;
      }
    }
```

Or even better. Find a way to reduce the complexity of this function!

Comment (quanho) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: Refactored!

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js [hunk 2]
```diff
@@ -32,9 +32,29 @@ function toBackground(label) {
...
-    : !(rect.width < textMetrics.width || rect.height < textMetrics.height);
+function isFitWidth(rect, label, rotate, ignoreCheck) {
+  return ignoreCheck || (rotate ? rect.width >= label.height : rect.width >= label.width);
+}
+
+function isFitHeight(rect, label, rotate, ignoreCheck) {
+  return ignoreCheck || (rotate ? rect.height >= label.width : rect.height >= label.height);
+}
+
+function isTextFitRect(orientation, rect, label, fitsHorizontally, overflow) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: So what does isTextFitRect mean?

Comment (cbt1) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: And how does it differ from isTextInRect?

Comment (quanho) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: isTextFitRect  takes different types of parameter than isTextInRect. In core, they will use the same things (isFitWidth and isFitHeight).

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js [hunk 3]
```diff
@@ -32,9 +32,29 @@ function toBackground(label) {
...
+  let fitWidth;
+  let fitHeight;
+  if (orientation === 'v') {
+    fitWidth = fitsHorizontally || isFitWidth(rect, label, !fitsHorizontally, overflow);
```
Comment (cbt1) [inline packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: I don't think it adds any value to pass in the `overflow` variable to "isFitXXXX". It would be easier to read if it parts of this statement instead. Such that `fitWidth = fitsHorizontally || overflow || isFitWidth(rect, label, !fitsHorizontally);`

It would also skip one computation when the overflow parameter is not passed in to the function.

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js (hunk 2 above)
Comment (cbt1) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: I do realize that naming is very hard, but `isTextFitRect, isFitWidth and isFitHeight` doesn't really mean anything. So I suggest you try and be a bit more descriptive and name them after their purpose in plain english.

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js (hunk 3 above)
Comment (quanho) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: Fixed!

File: packages/picasso.js/src/core/chart-components/labels/strategies/bars.js (hunk 2 above)
Comment (quanho) [reply packages/picasso.js/src/core/chart-components/labels/strategies/bars.js]: Updated!
